File: nexus-sdk/typescript/src/wallet.ts
/**
 * NexusChain SDK 钱包管理模块（v2.2.0 补真，2026-09-01）。
 *
 * 查询走 nexus-core JSON-RPC 真实信封；交易提交走 nexus-wallet-service
 * HTTP（core 无 nexus_sendRawTransaction——签名与密钥管控是 wallet-service
 * 的架构职责，SDK 不持私钥）。地址校验为纯本地实现（Base58 + 25 字节 +
 * keccak 双哈希校验尾，对齐 Java KeystoreAction.verifyAddress）。
 */

import type { RpcClient } from './rpc';
import type { WalletInfo } from './types';
import { validateAddress } from './address';
import axios from 'axios';

/** 钱包管理器 */
export class WalletManager {
  private readonly rpcClient: RpcClient;
  private readonly network: string;
  private readonly walletServiceUrl?: string;
  private readonly apiKey?: string;

  constructor(
    rpcClient: RpcClient,
    network: string,
    walletServiceUrl?: string,
    apiKey?: string,
  ) {
    this.rpcClient = rpcClient;
    this.network = network;
    this.walletServiceUrl = walletServiceUrl;
    this.apiKey = apiKey;
  }

  /**
   * 创建新钱包。
   *
   * 架构决策（非缺陷）：密钥生成由 wallet-service 执行（KMS、轮换、审计
   * 策略集中管控），SDK 不在客户端生成私钥。调用 wallet-service 的
   * /api/v1/wallets 端点。
   */
  async create(): Promise<WalletInfo> {
    throw new Error(
      'WalletManager.create: key generation is wallet-service\'s domain ' +
        '(KMS/rotation/audit). POST /api/v1/wallets on nexus-wallet-service.',
    );
  }

  /** NEX 余额（最小单位）。信封：nexus_getBalance → {"balance":"<decimal>"}。 */
  async getBalance(address: string): Promise<string> {
    const result = (await this.rpcClient.call('nexus_getBalance', [address, 'latest'])) as unknown;
    if (typeof result === 'object' && result !== null) {
      const obj = result as Record<string, unknown>;
      if (obj.balance !== undefined) return String(obj.balance);
    }
    return String(result);
  }

  /** 下一 nonce。信封：nexus_getTransactionCount → {"count":N}。 */
  async getNonce(address: string): Promise<number> {
    const result = (await this.rpcClient.call('nexus_getTransactionCount', [address])) as unknown;
    if (typeof result === 'object' && result !== null) {
      const c = (result as Record<string, unknown>).count;
      if (c !== undefined) return Number(c);
    }
    throw new Error(`unexpected count envelope: ${typeof result}`);
  }

  /** 按地址查交易（nexus_getTransactionsByAddress → dict[]）。 */
  async getTransactionsByAddress(address: string, limit = 20): Promise<Record<string, unknown>[]> {
    const result = (await this.rpcClient.call('nexus_getTransactionsByAddress', [address, limit])) as unknown;
    if (Array.isArray(result)) return result as Record<string, unknown>[];
    throw new Error(`unexpected transaction list envelope: ${typeof result}`);
  }

  /** 本地地址校验（Base58 + 25 字节 + keccak 双哈希校验尾，不联网）。 */
  validateAddress(address: string): boolean {
    return validateAddress(address);
  }

  /**
   * 通过 wallet-service 签名并提交转账，返回交易哈希。
   *
   * 需要 NexusChainConfig.walletServiceUrl；本地先做地址格式校验。
   */
  async submitTransfer(from: string, to: string, amount: string): Promise<string> {
    if (!this.walletServiceUrl) {
      throw new Error(
        'walletServiceUrl is required for submitTransfer ' +
          '(submission goes through nexus-wallet-service, not core JSON-RPC)',
      );
    }
    if (!validateAddress(from)) throw new Error(`invalid from address: ${from}`);
    if (!validateAddress(to)) throw new Error(`invalid to address: ${to}`);

    const resp = await axios.post(
      `${this.walletServiceUrl.replace(/\/$/, '')}/api/v1/transfers`,
      { from, to, amount, token: 'NEX' },
      {
        headers: {
          'Content-Type': 'application/json',
          ...(this.apiKey ? { Authorization: `Bearer ${this.apiKey}` } : {}),
        },
      },
    );
    const body = resp.data as Record<string, unknown>;
    const candidates = [
      body.txHash,
      body.hash,
      (body.data as Record<string, unknown> | undefined)?.txHash,
    ];
    for (const c of candidates) {
      if (typeof c === 'string' && c) return c;
    }
    throw new Error(`wallet-service response missing tx hash: ${JSON.stringify(body)}`);
  }
}

File: nexus-sdk/go/nexus/wallet.go
package nexus

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"math/big"
	"net/http"

	"golang.org/x/crypto/sha3"
)

// WalletManager provides wallet queries against nexus-core and transaction
// submission through nexus-wallet-service.
//
// Key generation / signing deliberately live in wallet-service (KMS, rotation,
// audit) — the SDK never holds private keys (see package doc).
type WalletManager struct {
	client *Client
}

// GetBalance returns the NEX balance in minimal units (wei).
//
// Envelope (JsonRpcController.doGetBalance): {"balance": "<decimal string>"}.
func (w *WalletManager) GetBalance(address string) (*big.Int, error) {
	result, err := w.client.RPCCall("nexus_getBalance", address)
	if err != nil {
		return nil, fmt.Errorf("nexus: get balance: %w", err)
	}
	m, ok := result.(map[string]interface{})
	if !ok {
		return nil, fmt.Errorf("nexus: unexpected balance envelope: %T", result)
	}
	b, ok := new(big.Int).SetString(fmt.Sprintf("%v", m["balance"]), 10)
	if !ok {
		return nil, fmt.Errorf("nexus: balance not a decimal string: %v", m["balance"])
	}
	return b, nil
}

// GetNonce returns the next nonce for an address.
//
// Envelope (doGetTransactionCount): {"count": <numeric>}.
func (w *WalletManager) GetNonce(address string) (int64, error) {
	result, err := w.client.RPCCall("nexus_getTransactionCount", address)
	if err != nil {
		return 0, fmt.Errorf("nexus: get nonce: %w", err)
	}
	m, ok := result.(map[string]interface{})
	if !ok {
		return 0, fmt.Errorf("nexus: unexpected count envelope: %T", result)
	}
	return toInt64(m["count"])
}

// GetTransactionsByAddress returns transactions touching an address
// (nexus_getTransactionsByAddress → RpcTransaction[]).
func (w *WalletManager) GetTransactionsByAddress(address string, limit int) ([]map[string]interface{}, error) {
	result, err := w.client.RPCCall("nexus_getTransactionsByAddress", address, limit)
	if err != nil {
		return nil, fmt.Errorf("nexus: get transactions: %w", err)
	}
	arr, ok := result.([]interface{})
	if !ok {
		return nil, fmt.Errorf("nexus: unexpected transaction list envelope: %T", result)
	}
	out := make([]map[string]interface{}, 0, len(arr))
	for _, e := range arr {
		if m, ok := e.(map[string]interface{}); ok {
			out = append(out, m)
		}
	}
	return out, nil
}

// SubmitTransfer signs and submits a transfer through nexus-wallet-service.
//
// core's JSON-RPC has no nexus_sendRawTransaction (deliberate: transaction
// submission is wallet-service's domain). The SDK posts a structured request;
// wallet-service enforces KMS/signing/audit and returns the tx hash.
type SubmitTransferRequest struct {
	From   string   `json:"from"`
	To     string   `json:"to"`
	Amount *big.Int `json:"amount"`
	Token  string   `json:"token,omitempty"`
}

// SubmitTransfer posts the transfer to wallet-service /api/v1/transfers and
// returns the resulting transaction hash.
func (w *WalletManager) SubmitTransfer(req SubmitTransferRequest) (string, error) {
	if w.client.config.WalletServiceURL == "" {
		return "", fmt.Errorf("nexus: Config.WalletServiceURL is required for SubmitTransfer " +
			"(transaction submission goes through nexus-wallet-service, not core JSON-RPC)")
	}
	if !ValidateAddress(req.From) || !ValidateAddress(req.To) {
		return "", fmt.Errorf("nexus: invalid address (from=%q to=%q)", req.From, req.To)
	}
	body, err := json.Marshal(map[string]interface{}{
		"from":   req.From,
		"to":     req.To,
		"amount": req.Amount.String(),
		"token":  "NEX",
	})
	if err != nil {
		return "", err
	}
	httpReq, err := http.NewRequest("POST",
		w.client.config.WalletServiceURL+"/api/v1/transfers", bytes.NewReader(body))
	if err != nil {
		return "", err
	}
	httpReq.Header.Set("Content-Type", "application/json")
	if w.client.config.APIKey != "" {
		httpReq.Header.Set("Authorization", "Bearer "+w.client.config.APIKey)
	}
	resp, err := w.client.httpClient.Do(httpReq)
	if err != nil {
		return "", fmt.Errorf("nexus: wallet-service request failed: %w", err)
	}
	defer resp.Body.Close()
	respBody, _ := io.ReadAll(resp.Body)
	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusCreated {
		return "", fmt.Errorf("nexus: wallet-service http %d: %s", resp.StatusCode, truncate(respBody, 200))
	}
	var envelope struct {
		TxHash string `json:"txHash"`
		Hash   string `json:"hash"`
		Data   struct {
			TxHash string `json:"txHash"`
		} `json:"data"`
	}
	if err := json.Unmarshal(respBody, &envelope); err != nil {
		return "", fmt.Errorf("nexus: decode wallet-service response: %w", err)
	}
	if h := envelope.TxHash; h != "" {
		return h, nil
	}
	if h := envelope.Hash; h != "" {
		return h, nil
	}
	if h := envelope.Data.TxHash; h != "" {
		return h, nil
	}
	return "", fmt.Errorf("nexus: wallet-service response missing tx hash: %s", truncate(respBody, 200))
}

// ValidateAddress checks NexusChain address format locally (no network):
// Base58 decodable to 25 bytes (1 version byte 0x00 + 20 pubkey-hash + 4
// checksum), checksum = first 4 bytes of keccak256(keccak256(pubkeyHash))
// where pubkeyHash = decoded[1:21] (version byte skipped). Mirrors Java
// KeystoreAction.verifyAddress: r5=decode(addr) is 25 bytes; r2=r5[0:21] for
// pubkeyHash extraction; checksum compares keccak²(pubkeyHash) with r5[-4:].
func ValidateAddress(address string) bool {
	decoded, err := base58Decode(address)
	if err != nil || len(decoded) != 25 {
		return false
	}
	pubkeyHash := decoded[1:21]
	checksum := decoded[len(decoded)-4:]
	h := keccak256(keccak256(pubkeyHash))
	for i := 0; i < 4; i++ {
		if h[i] != checksum[i] {
			return false
		}
	}
	return true
}

// base58Decode implements the Bitcoin-alphabet Base58 decode (no dependency).
func base58Decode(s string) ([]byte, error) {
	alphabet := "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"
	index := make(map[byte]int, 58)
	for i := 0; i < len(alphabet); i++ {
		index[alphabet[i]] = i
	}
	zeros := 0
	for zeros < len(s) && s[zeros] == '1' {
		zeros++
	}
	// big-endian base-58 → big integer
	n := new(big.Int)
	radix := big.NewInt(58)
	for i := 0; i < len(s); i++ {
		v, ok := index[s[i]]
		if !ok {
			return nil, fmt.Errorf("invalid base58 character %q", s[i])
		}
		n.Mul(n, radix)
		n.Add(n, big.NewInt(int64(v)))
	}
	out := make([]byte, zeros)
	body := n.Bytes()
	if n.Sign() == 0 {
		body = nil
	}
	out = append(out, body...)
	return out, nil
}

// keccak256 uses SHA3-256 with the original Keccak padding (NOT NIST SHA3) —
// the checksum flavor NexusChain addresses use (mirrors Java SHA3Utility.keccak256).
func keccak256(data []byte) []byte {
	h := sha3.NewLegacyKeccak256()
	h.Write(data)
	return h.Sum(nil)
}
